src/charm.py

```diff
@@ -198,6 +214,36 @@ def _on_leader_elected(self, _):
                 password = generate_random_password(PASSWORD_LENGTH)
                 peer_data[required_password] = password
 
+        # If this node was elected a leader due to a prior leader unit being down scaled
+        if self._is_peer_data_set and self._is_cluster_initialized:
+            self.unit.status = MaintenanceStatus("Removing unreachable instances")
+
+            # Remove unreachable instances from the cluster
+            self._mysql.remove_instances_not_online()
+
+            # Since we do not have name of the prior leader unit, use planned units
+            # to remove units with index >= planned units from the allowlist
+            planned_units = self.app.planned_units()
+            allowlist = set(self._peers.data[self.app].get("allowlist").split(","))
+            new_allowlist = set(allowlist)
+
+            for fqdn in allowlist:
```

Comment:
you might be able to simplify this with some set math like how Mykola did when he handled this issue in the [k8s charm](https://github.com/delgod/mongodb-k8s-operator/blob/f7e8690b7e7552e226cc57ccf51ba24b562243af/src/charm.py#L177-L180) it might simplify the logic, but it would require that you keep a list of the current set members. 🤷

Reply:
We actually may be handling allowlists incorrectly (we're only setting them on the juju leader unit, whereas we might need to set them in all units). Will followup after speaking with Mykola